File: src/main/java/core/scm/Procedure.java
package core.scm;

import core.environment.Environment;
import core.procedures.AFn;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Map;

/* Lambda */
public class Procedure extends AFn {

  private static final Symbol[] EMPTY = new Symbol[0];

  private String name;

  /* Array of arguments the procedure expects */
  private final Symbol[] args;

  /* Body form of the procedure */
  private final Object body;

  /* Is body a constant? If it is, then no need to evaluate it */
  private final boolean isBodyConst;

  /* Lexical environment */
  private Environment localEnvironment = null;

  public Procedure(String name, List<Symbol> args, Object body, Environment localEnvironment, boolean isVariadic) {
    this(name, args.toArray(EMPTY), body, localEnvironment, isVariadic);
  }

  public Procedure(String name, Symbol[] args, Object body, Environment localEnvironment, boolean isVariadic) {
    this.name = name;
    this.args = args == null ? EMPTY : args;
    this.body = body;
    this.isBodyConst = isConst(body);
    this.localEnvironment = localEnvironment;
    if (isVariadic) {
      /* Do not count rest arg */
      this.setMinArgs(this.args.length - 1);
    } else {
      this.setMinArgs(this.args.length);
      this.setMaxArgs(this.args.length);
    }
  }

  private static boolean isConst(Object obj) {
    return !((obj instanceof Symbol) || (obj instanceof Collection) || (obj instanceof Map));
  }

  private Symbol[] getArgs() {
    return args;
  }

  public void setName(String name) {
    this.name = name;
  }

  private Environment bindArgs(Object... values) {
    /* Evaluate mandatory params and put values into new local environment */
    Environment env = new Environment(values.length, this.localEnvironment);
    Symbol[] args = getArgs();
    for (int i = 0; i < getMinArgs(); i++) {
      env.put(args[i], values[i]);
    }
    /* If it is a variadic function, then evaluate rest param */
    if (getMinArgs() != getMaxArgs()) {
      /* Optional params: pass them as a list bound to the last param.
       * Everything AFTER mandatory params goes to that list. */
      env.put(args[minArgs()], Arrays.asList(Arrays.copyOfRange(values, minArgs(), values.length)));
    }
    return env;
  }

  @Override
  public Object apply0() {
    if (isBodyConst) {
      return body;
    }
    return new Thunk(body, new Environment(0, this.localEnvironment));
  }

  @Override
  public Object apply1(Object arg1) {
    if (isBodyConst) {
      return body;
    }
    Environment environment = new Environment(1, this.localEnvironment);
    environment.put(args[0], arg1);
    return new Thunk(body, environment);
  }

  @Override
  public Object apply2(Object arg1, Object arg2) {
    if (isBodyConst) {
      return body;
    }
    Environment environment = new Environment(2, this.localEnvironment);
    environment.put(args[0], arg1);
    environment.put(args[1], arg2);
    return new Thunk(body, environment);
  }

  @Override
  public Object apply3(Object arg1, Object arg2, Object arg3) {
    if (isBodyConst) {
      return body;
    }
    Environment environment = new Environment(3, this.localEnvironment);
    environment.put(args[0], arg1);
    environment.put(args[1], arg2);
    environment.put(args[2], arg3);
    return new Thunk(body, environment);
  }

  @Override
  public Object apply4(Object arg1, Object arg2, Object arg3, Object arg4) {
    if (isBodyConst) {
      return body;
    }
    Environment environment = new Environment(4, this.localEnvironment);
    environment.put(args[0], arg1);
    environment.put(args[1], arg2);
    environment.put(args[2], arg3);
    environment.put(args[3], arg4);
    return new Thunk(body, environment);
  }

  @Override
  public Object apply(Object... args) {
    return isBodyConst ? body : new Thunk(body, bindArgs(args));
  }

  @Override
  public String getName() {
    return name;
  }
}

File: src/main/java/core/scm/Procedure.kt
package core.scm

import core.environment.Environment
import core.procedures.AFn

/* Lambda */
class Procedure(override var name: String,
                /* Array of arguments the procedure expects */
                private val args: Array<Symbol?>,
                /* Body form of the procedure */
                private val body: Any,
                /* Lexical environment */
                private val localEnvironment: Environment,
                isVariadic: Boolean) : AFn<Any?, Any?>() {

    /* Is body a constant? If it is, then no need to evaluate it */
    private val isBodyConst: Boolean = body !is Symbol && body !is Collection<*> && body !is Map<*, *>

    init {
        if (isVariadic) {
            /* Do not count `rest` arg */
            this.minArgs = this.args.size - 1
        } else {
            this.minArgs = this.args.size
            this.maxArgs = this.args.size
        }
    }

    override operator fun invoke() = when {
        isBodyConst -> body
        else        -> Thunk(body, Environment(0, localEnvironment))
    }

    override operator fun invoke(arg: Any?) = when {
        isBodyConst -> body
        else -> Thunk(body, Environment(1, localEnvironment).apply { put(args[0], arg) })
    }

    override operator fun invoke(arg1: Any?, arg2: Any?) = when {
        isBodyConst -> body
        else -> Thunk(body, Environment(2, localEnvironment).apply {
            put(args[0], arg1)
            put(args[1], arg2)
        })
    }

    override operator fun invoke(arg1: Any?, arg2: Any?, arg3: Any?) = when {
        isBodyConst -> body
        else -> Thunk(body, Environment(3, localEnvironment).apply {
            put(args[0], arg1)
            put(args[1], arg2)
            put(args[2], arg3)
        })
    }

    override operator fun invoke(arg1: Any?, arg2: Any?, arg3: Any?, arg4: Any?) = when {
        isBodyConst -> body
        else -> Thunk(body, Environment(4, localEnvironment).apply {
            put(args[0], arg1)
            put(args[1], arg2)
            put(args[2], arg3)
            put(args[3], arg4)
        })
    }

    override operator fun invoke(args: Array<out Any?>) = if (isBodyConst) body else Thunk(body, bindArgs(args))

    private fun bindArgs(values: Array<out Any?>) = Environment(values.size, localEnvironment).apply {
        /* Evaluate mandatory params and put values into new local environment */
        for (i in 0 until minArgs) {
            put(args[i], values[i])
        }
        if (minArgs != maxArgs) {
            /* Optional params: pass them as a list bound to the last param.
             * Everything AFTER mandatory params goes to that list. */
            put(args[minArgs], values.copyOfRange(minArgs, values.size).asList())
        }
    }

    /* Lambdas have no arg type information, hence nothing to check */
    override fun checkArgs(args: Array<out Any?>) = Unit
}
